File: Add Two Numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *addTwoNumbers(ListNode *l1, ListNode *l2) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
        if (l1 == NULL) {
            return l2;
        } else if (l2 == NULL) {
            return l1;
        }
        
        int carry = 0;
        ListNode *head = new ListNode(-1);
        ListNode *pre = head;
        while (l1 && l2) {
            int sum = l1->val + l2->val + carry;
            carry = sum / 10;
            sum = sum % 10;
            ListNode *cur = new ListNode(sum);
            pre->next = cur;
            pre = cur;
            l1 = l1->next;
            l2 = l2->next;
        }
        
        if (l2 != NULL) {
            l1 = l2;
        }
        
        while (l1 != NULL) {
            int sum = l1->val + carry;
            carry = sum / 10;
            sum = sum % 10;
            ListNode *cur = new ListNode(sum);
            pre->next = cur;
            pre = cur;
            l1 = l1->next;
        }
        if (carry == 1) {
            ListNode *cur = new ListNode(1);
            pre->next = cur;
            pre = cur;
        }
        pre->next = NULL;
        
        return head->next;
    }
};

File: Add Two Numbers.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode AddTwoNumbers(ListNode l1, ListNode l2) {
        if (l1 == null) {
            return l2;
        }  
        if (l2 == null) {
            return l1;
        }
        
        var carry = 0;
        var head = new ListNode(-1);
        var pre = head;
        while (l1 != null && l2 != null) {
            var sum = l1.val + l2.val + carry;
            carry = sum / 10;
            var cur = new ListNode(sum%10);
            pre.next = cur;
            pre = cur;
            
            l1 = l1.next;
            l2 = l2.next;
        }
        
        if (l2 != null) {
            l1 = l2;
        }
        
        while (l1 != null) {
            var sum = l1.val + carry;
            carry = sum / 10;
            var cur = new ListNode(sum%10);
            pre.next = cur;
            pre = cur;
            
            l1 = l1.next;
        }
        
        if (carry == 1) {
            var cur = new ListNode(1);
            pre.next = cur;
        }

        return head.next;
    }
}




